One of the errors:

```
[ERROR]   ServletITCase>ServletTestCase.testWebjars:63 1 expectation failed.
Expected status code <200> but was <404>.
```

--- a/core/deployment/src/main/java/io/quarkus/deployment/builditem/BytecodeTransformerBuildItem.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/builditem/BytecodeTransformerBuildItem.java
@@ -1,5 +1,6 @@
 package io.quarkus.deployment.builditem;
 
+import java.util.Set;
 import java.util.function.BiFunction;
 
 import org.objectweb.asm.ClassVisitor;
@@ -18,16 +19,38 @@ public final class BytecodeTransformerBuildItem extends MultiBuildItem {
     final String classToTransform;
     final BiFunction<String, ClassVisitor, ClassVisitor> visitorFunction;
 
+    /**
+     * A set of class names that need to be present in the const pool for the transformation to happen. These
+     * need to be in JVM internal format.
+     *
+     * The transformation is only applied if at least one of the entries in the const pool is present
+     *
+     * Note that this is an optimisation, and if another transformer is transforming the class anyway then
+     * this transformer will always be applied.
+     */
+    final Set<String> requireConstPoolEntry;
+
     public BytecodeTransformerBuildItem(String classToTransform,
             BiFunction<String, ClassVisitor, ClassVisitor> visitorFunction) {
-        this(false, classToTransform, visitorFunction);
+        this(classToTransform, visitorFunction, null);
+    }
+
+    public BytecodeTransformerBuildItem(String classToTransform,
+            BiFunction<String, ClassVisitor, ClassVisitor> visitorFunction, Set<String> requireConstPoolEntry) {
+        this(false, classToTransform, visitorFunction, requireConstPoolEntry);
     }
 
     public BytecodeTransformerBuildItem(boolean eager, String classToTransform,
             BiFunction<String, ClassVisitor, ClassVisitor> visitorFunction) {
+        this(eager, classToTransform, visitorFunction, null);
+    }
+
+    public BytecodeTransformerBuildItem(boolean eager, String classToTransform,
+            BiFunction<String, ClassVisitor, ClassVisitor> visitorFunction, Set<String> requireConstPoolEntry) {
         this.eager = eager;
         this.classToTransform = classToTransform;
         this.visitorFunction = visitorFunction;
+        this.requireConstPoolEntry = requireConstPoolEntry;
     }
 
     public String getClassToTransform() {
@@ -36,6 +59,10 @@ public final class BytecodeTransformerBuildItem extends MultiBuildItem {
 
     public BiFunction<String, ClassVisitor, ClassVisitor> getVisitorFunction() {
         return visitorFunction;
+    }
+
+    public Set<String> getRequireConstPoolEntry() {
+        return requireConstPoolEntry;
     }
 
     public boolean isEager() {

--- a/core/deployment/src/main/java/io/quarkus/deployment/index/ConstPoolPredicate.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/index/ConstPoolPredicate.java
@@ -1,0 +1,4 @@
+package io.quarkus.deployment.index;
+
+public class ConstPoolPredicate {
+}

--- a/core/deployment/src/main/java/io/quarkus/deployment/index/ConstPoolScanner.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/index/ConstPoolScanner.java
@@ -1,0 +1,92 @@
+package io.quarkus.deployment.index;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Set;
+
+public class ConstPoolScanner {
+
+    static final int CONSTANT_UTF8_TAG = 1;
+    static final int CONSTANT_INTEGER_TAG = 3;
+    static final int CONSTANT_FLOAT_TAG = 4;
+    static final int CONSTANT_LONG_TAG = 5;
+    static final int CONSTANT_DOUBLE_TAG = 6;
+    static final int CONSTANT_CLASS_TAG = 7;
+    static final int CONSTANT_STRING_TAG = 8;
+    static final int CONSTANT_FIELDREF_TAG = 9;
+    static final int CONSTANT_METHODREF_TAG = 10;
+    static final int CONSTANT_INTERFACE_METHODREF_TAG = 11;
+    static final int CONSTANT_NAME_AND_TYPE_TAG = 12;
+    static final int CONSTANT_METHOD_HANDLE_TAG = 15;
+    static final int CONSTANT_METHOD_TYPE_TAG = 16;
+    static final int CONSTANT_DYNAMIC_TAG = 17;
+    static final int CONSTANT_INVOKE_DYNAMIC_TAG = 18;
+    static final int CONSTANT_MODULE_TAG = 19;
+    static final int CONSTANT_PACKAGE_TAG = 20;
+
+    //TODO: at the moment this only looks for the class name, it does not make sure it is used
+    //by a class tag. In practice this is likely fine, especially as this is just used for optimisations.
+    public static boolean constPoolEntryPresent(byte[] classBody, Set<String> namesToLookFor) {
+        ByteBuffer data = ByteBuffer.wrap(classBody);
+        if (data.getInt() != 0xCAFEBABE) {
+            return false; //not a class file
+        }
+        data.getShort();//major
+        data.getShort();//minor
+        int constantPoolCount = data.getShort();
+        int currentCpInfoIndex = 1;
+        while (currentCpInfoIndex < constantPoolCount) {
+            currentCpInfoIndex++;
+            int cpInfoSize;
+            switch (data.get()) {
+                case CONSTANT_FIELDREF_TAG:
+                case CONSTANT_METHODREF_TAG:
+                case CONSTANT_INTERFACE_METHODREF_TAG:
+                case CONSTANT_INTEGER_TAG:
+                case CONSTANT_FLOAT_TAG:
+                case CONSTANT_NAME_AND_TYPE_TAG:
+                    cpInfoSize = 4;
+                    break;
+                case CONSTANT_DYNAMIC_TAG:
+                    cpInfoSize = 4;
+                    break;
+                case CONSTANT_INVOKE_DYNAMIC_TAG:
+                    cpInfoSize = 4;
+                    break;
+                case CONSTANT_LONG_TAG:
+                case CONSTANT_DOUBLE_TAG:
+                    cpInfoSize = 8;
+                    currentCpInfoIndex++;
+                    break;
+                case CONSTANT_UTF8_TAG:
+                    int strLength = 0xFFFF & data.getShort();
+                    cpInfoSize = 0;
+                    byte[] str = new byte[strLength];
+                    data.get(str);
+                    if (namesToLookFor.contains(new String(str, StandardCharsets.UTF_8))) {
+                        return true;
+                    }
+                    break;
+                case CONSTANT_METHOD_HANDLE_TAG:
+                    cpInfoSize = 3;
+                    break;
+                case CONSTANT_CLASS_TAG:
+                case CONSTANT_STRING_TAG:
+                case CONSTANT_METHOD_TYPE_TAG:
+                case CONSTANT_PACKAGE_TAG:
+                case CONSTANT_MODULE_TAG:
+                    cpInfoSize = 2;
+                    break;
+                default:
+                    throw new IllegalArgumentException();
+            }
+            for (int j = 0; j < cpInfoSize; ++j) {
+                data.get();
+            }
+        }
+
+        return false;
+
+    }
+
+}

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/ClassTransformingBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/ClassTransformingBuildStep.java
@@ -29,6 +29,7 @@ import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.builditem.ApplicationArchivesBuildItem;
 import io.quarkus.deployment.builditem.BytecodeTransformerBuildItem;
 import io.quarkus.deployment.builditem.TransformedClassesBuildItem;
+import io.quarkus.deployment.index.ConstPoolScanner;
 
 public class ClassTransformingBuildStep {
 
@@ -42,9 +43,17 @@ public class ClassTransformingBuildStep {
         }
         final Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> bytecodeTransformers = new HashMap<>(
                 bytecodeTransformerBuildItems.size());
+        Set<String> noConstScanning = new HashSet<>();
+        Map<String, Set<String>> constScanning = new HashMap<>();
         for (BytecodeTransformerBuildItem i : bytecodeTransformerBuildItems) {
             bytecodeTransformers.computeIfAbsent(i.getClassToTransform(), (h) -> new ArrayList<>())
                     .add(i.getVisitorFunction());
+            if (i.getRequireConstPoolEntry() == null || i.getRequireConstPoolEntry().isEmpty()) {
+                noConstScanning.add(i.getClassToTransform());
+            } else {
+                constScanning.computeIfAbsent(i.getClassToTransform(), (s) -> new HashSet<>())
+                        .addAll(i.getRequireConstPoolEntry());
+            }
         }
         Map<String, Path> transformedToArchive = new ConcurrentHashMap<>();
         // now copy all the contents to the runner jar
@@ -57,6 +66,7 @@ public class ClassTransformingBuildStep {
             for (Map.Entry<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> entry : bytecodeTransformers
                     .entrySet()) {
                 String className = entry.getKey();
+                Set<String> constValues = constScanning.get(className);
                 ApplicationArchive archive = appArchives.containingArchive(entry.getKey());
                 if (archive != null) {
                     List<BiFunction<String, ClassVisitor, ClassVisitor>> visitors = entry.getValue();
@@ -73,7 +83,14 @@ public class ClassTransformingBuildStep {
                                         throw new RuntimeException(
                                                 "Can't process class files larger than Integer.MAX_VALUE bytes");
                                     }
-                                    ClassReader cr = new ClassReader(Files.readAllBytes(classFile));
+                                    byte[] classData = Files.readAllBytes(classFile);
+
+                                    if (constValues != null && !noConstScanning.contains(className)) {
+                                        if (!ConstPoolScanner.constPoolEntryPresent(classData, constValues)) {
+                                            return null;
+                                        }
+                                    }
+                                    ClassReader cr = new ClassReader(classData);
                                     ClassWriter writer = new QuarkusClassWriter(cr,
                                             ClassWriter.COMPUTE_FRAMES | ClassWriter.COMPUTE_MAXS);
                                     ClassVisitor visitor = writer;
@@ -102,8 +119,10 @@ public class ClassTransformingBuildStep {
         if (!transformed.isEmpty()) {
             for (Future<TransformedClassesBuildItem.TransformedClass> i : transformed) {
                 final TransformedClassesBuildItem.TransformedClass res = i.get();
-                transformedClassesByJar.computeIfAbsent(transformedToArchive.get(res.getFileName()), (a) -> new HashSet<>())
-                        .add(res);
+                if (res != null) {
+                    transformedClassesByJar.computeIfAbsent(transformedToArchive.get(res.getFileName()), (a) -> new HashSet<>())
+                            .add(res);
+                }
             }
         }
         return new TransformedClassesBuildItem(transformedClassesByJar);

--- a/core/deployment/src/main/java/io/quarkus/runner/bootstrap/StartupActionImpl.java
+++ b/core/deployment/src/main/java/io/quarkus/runner/bootstrap/StartupActionImpl.java
@@ -19,6 +19,7 @@ import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.function.BiFunction;
 import java.util.function.Consumer;
+import java.util.function.Predicate;
 
 import org.jboss.logging.Logger;
 import org.objectweb.asm.ClassVisitor;
@@ -35,6 +36,7 @@ import io.quarkus.deployment.builditem.GeneratedClassBuildItem;
 import io.quarkus.deployment.builditem.GeneratedResourceBuildItem;
 import io.quarkus.deployment.builditem.MainClassBuildItem;
 import io.quarkus.deployment.configuration.RunTimeConfigurationGenerator;
+import io.quarkus.deployment.index.ConstPoolScanner;
 import io.quarkus.dev.appstate.ApplicationStateNotification;
 import io.quarkus.runtime.Quarkus;
 
@@ -51,8 +53,9 @@ public class StartupActionImpl implements StartupAction {
         this.curatedApplication = curatedApplication;
         this.buildResult = buildResult;
         Set<String> eagerClasses = new HashSet<>();
+        Map<String, Predicate<byte[]>> transformerPredicates = new HashMap<>();
         Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> bytecodeTransformers = extractTransformers(
-                eagerClasses);
+                eagerClasses, transformerPredicates);
         QuarkusClassLoader baseClassLoader = curatedApplication.getBaseRuntimeClassLoader();
         QuarkusClassLoader runtimeClassLoader;
 
@@ -60,15 +63,16 @@ public class StartupActionImpl implements StartupAction {
         //test mode only has a single class loader, while dev uses a disposable runtime class loader
         //that is discarded between restarts
         if (curatedApplication.getQuarkusBootstrap().getMode() == QuarkusBootstrap.Mode.DEV) {
-            baseClassLoader.reset(extractGeneratedResources(false), bytecodeTransformers, deploymentClassLoader);
+            baseClassLoader.reset(extractGeneratedResources(false), bytecodeTransformers, transformerPredicates,
+                    deploymentClassLoader);
             runtimeClassLoader = curatedApplication.createRuntimeClassLoader(baseClassLoader,
-                    bytecodeTransformers,
+                    bytecodeTransformers, transformerPredicates,
                     deploymentClassLoader, extractGeneratedResources(true));
         } else {
             Map<String, byte[]> resources = new HashMap<>();
             resources.putAll(extractGeneratedResources(false));
             resources.putAll(extractGeneratedResources(true));
-            baseClassLoader.reset(resources, bytecodeTransformers, deploymentClassLoader);
+            baseClassLoader.reset(resources, bytecodeTransformers, transformerPredicates, deploymentClassLoader);
             runtimeClassLoader = baseClassLoader;
         }
         this.runtimeClassLoader = runtimeClassLoader;
@@ -257,8 +261,11 @@ public class StartupActionImpl implements StartupAction {
         return runtimeClassLoader;
     }
 
-    private Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> extractTransformers(Set<String> eagerClasses) {
+    private Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> extractTransformers(Set<String> eagerClasses,
+            Map<String, Predicate<byte[]>> transformerPredicates) {
         Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> bytecodeTransformers = new HashMap<>();
+        Set<String> noConstScanning = new HashSet<>();
+        Map<String, Set<String>> constScanning = new HashMap<>();
         List<BytecodeTransformerBuildItem> transformers = buildResult.consumeMulti(BytecodeTransformerBuildItem.class);
         for (BytecodeTransformerBuildItem i : transformers) {
             List<BiFunction<String, ClassVisitor, ClassVisitor>> list = bytecodeTransformers.get(i.getClassToTransform());
@@ -269,6 +276,23 @@ public class StartupActionImpl implements StartupAction {
             if (i.isEager()) {
                 eagerClasses.add(i.getClassToTransform());
             }
+            if (i.getRequireConstPoolEntry() == null || i.getRequireConstPoolEntry().isEmpty()) {
+                noConstScanning.add(i.getClassToTransform());
+            } else {
+                constScanning.computeIfAbsent(i.getClassToTransform(), (s) -> new HashSet<>())
+                        .addAll(i.getRequireConstPoolEntry());
+            }
+        }
+        for (String i : noConstScanning) {
+            constScanning.remove(i);
+        }
+        for (Map.Entry<String, Set<String>> entry : constScanning.entrySet()) {
+            transformerPredicates.put(entry.getKey(), new Predicate<byte[]>() {
+                @Override
+                public boolean test(byte[] bytes) {
+                    return ConstPoolScanner.constPoolEntryPresent(bytes, entry.getValue());
+                }
+            });
         }
         return bytecodeTransformers;
     }

--- a/docs/src/main/asciidoc/hibernate-orm-panache.adoc
+++ b/docs/src/main/asciidoc/hibernate-orm-panache.adoc
@@ -405,7 +405,7 @@ NOTE: The `stream` methods require a transaction to work.
 
 NOTE: The rest of the documentation show usages based on the active record pattern only,
 but keep in mind that they can be performed with the repository pattern as well.
-The repository pattern examples have been omitted for brevity. 
+The repository pattern examples have been omitted for brevity.
 
 == Advanced Query
 
@@ -944,7 +944,9 @@ That's all there is to it: with Panache, Hibernate ORM has never looked so trim 
 == Defining entities in external projects or jars
 
 Hibernate ORM with Panache relies on compile-time bytecode enhancements to your entities.
-If you define your entities in the same project where you build your Quarkus application, everything will work fine.
-If the entities come from external projects or jars, you can make sure that your jar is treated like a Quarkus application library
-by indexing it via Jandex, see link:cdi-reference#how-to-generate-a-jandex-index[How to Generate a Jandex Index] in the CDI guide.
-This will allow Quarkus to index and enhance your entities as if they were inside the current project.
+
+It attempts to identity archives with Panache entities (and consumers of Panache entities)
+by the presence of the marker file `META-INF/panache-archive.marker`. Panache includes an
+annotation processor that will automatically create this file in archives that depend on
+Panache (even indirectly). If you have disabled annotation processors you may need to create
+this file manually in some cases.

--- a/docs/src/main/asciidoc/mongodb-panache.adoc
+++ b/docs/src/main/asciidoc/mongodb-panache.adoc
@@ -1077,7 +1077,9 @@ That's all there is to it: with Panache, MongoDB has never looked so trim and ne
 == Defining entities in external projects or jars
 
 MongoDB with Panache relies on compile-time bytecode enhancements to your entities.
-If you define your entities in the same project where you build your Quarkus application, everything will work fine.
-If the entities come from external projects or jars, you can make sure that your jar is treated like a Quarkus application library
-by indexing it via Jandex, see link:cdi-reference#how-to-generate-a-jandex-index[How to Generate a Jandex Index] in the CDI guide.
-This will allow Quarkus to index and enhance your entities as if they were inside the current project.
+
+It attempts to identity archives with Panache entities (and consumers of Panache entities)
+by the presence of the marker file `META-INF/panache-archive.marker`. Panache includes an
+annotation processor that will automatically create this file in archives that depend on
+Panache (even indirectly). If you have disabled annotation processors you may need to create
+this file manually in some cases.

--- a/extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheHibernateResourceProcessor.java
+++ b/extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheHibernateResourceProcessor.java
@@ -21,11 +21,14 @@ import org.jboss.jandex.Type;
 
 import io.quarkus.arc.deployment.UnremovableBeanBuildItem;
 import io.quarkus.arc.deployment.ValidationPhaseBuildItem;
+import io.quarkus.bootstrap.classloading.ClassPathElement;
+import io.quarkus.bootstrap.classloading.QuarkusClassLoader;
 import io.quarkus.builder.BuildException;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.annotations.ExecutionTime;
 import io.quarkus.deployment.annotations.Record;
+import io.quarkus.deployment.builditem.AdditionalApplicationArchiveMarkerBuildItem;
 import io.quarkus.deployment.builditem.BytecodeTransformerBuildItem;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.builditem.FeatureBuildItem;
@@ -46,7 +49,7 @@ import io.quarkus.panache.common.deployment.PanacheMethodCustomizer;
 import io.quarkus.panache.common.deployment.PanacheMethodCustomizerBuildItem;
 import io.quarkus.panache.common.deployment.PanacheRepositoryEnhancer;
 
-public final class PanacheResourceProcessor {
+public final class PanacheHibernateResourceProcessor {
 
     static final DotName DOTNAME_PANACHE_REPOSITORY_BASE = DotName.createSimple(PanacheRepositoryBase.class.getName());
     private static final DotName DOTNAME_PANACHE_REPOSITORY = DotName.createSimple(PanacheRepository.class.getName());
@@ -58,6 +61,7 @@ public final class PanacheResourceProcessor {
     private static final DotName DOTNAME_NAMED_QUERY = DotName.createSimple(NamedQuery.class.getName());
     private static final DotName DOTNAME_NAMED_QUERIES = DotName.createSimple(NamedQueries.class.getName());
     private static final DotName DOTNAME_ID = DotName.createSimple(Id.class.getName());
+    protected static final String META_INF_PANACHE_ARCHIVE_MARKER = "META-INF/panache-archive.marker";
 
     @BuildStep
     FeatureBuildItem featureBuildItem() {
@@ -75,6 +79,11 @@ public final class PanacheResourceProcessor {
     @BuildStep
     UnremovableBeanBuildItem ensureBeanLookupAvailable() {
         return new UnremovableBeanBuildItem(new UnremovableBeanBuildItem.BeanTypeExclusion(DOTNAME_ENTITY_MANAGER));
+    }
+
+    @BuildStep
+    AdditionalApplicationArchiveMarkerBuildItem marker() {
+        return new AdditionalApplicationArchiveMarkerBuildItem(META_INF_PANACHE_ARCHIVE_MARKER);
     }
 
     @BuildStep
@@ -133,7 +142,9 @@ public final class PanacheResourceProcessor {
             if (modelClasses.add(classInfo.name().toString()))
                 modelEnhancer.collectFields(classInfo);
         }
+        Set<String> modelClassNamesInternal = new HashSet<>();
         for (String modelClass : modelClasses) {
+            modelClassNamesInternal.add(modelClass.replace(".", "/"));
             transformers.produce(new BytecodeTransformerBuildItem(true, modelClass, modelEnhancer));
 
             // lookup for `@NamedQuery` on the hierarchy and produce NamedQueryEntityClassBuildStep
@@ -148,10 +159,17 @@ public final class PanacheResourceProcessor {
         MetamodelInfo<EntityModel<EntityField>> modelInfo = modelEnhancer.getModelInfo();
         if (modelInfo.hasEntities()) {
             PanacheFieldAccessEnhancer panacheFieldAccessEnhancer = new PanacheFieldAccessEnhancer(modelInfo);
-            for (ClassInfo classInfo : index.getIndex().getKnownClasses()) {
-                String className = classInfo.name().toString();
-                if (!modelClasses.contains(className)) {
-                    transformers.produce(new BytecodeTransformerBuildItem(className, panacheFieldAccessEnhancer));
+            QuarkusClassLoader tccl = (QuarkusClassLoader) Thread.currentThread().getContextClassLoader();
+            List<ClassPathElement> archives = tccl.getElementsWithResource(META_INF_PANACHE_ARCHIVE_MARKER);
+            for (ClassPathElement i : archives) {
+                for (String res : i.getProvidedResources()) {
+                    if (res.endsWith(".class")) {
+                        String cn = res.replace("/", ".").substring(0, res.length() - 6);
+                        if (!modelClasses.contains(cn)) {
+                            transformers.produce(
+                                    new BytecodeTransformerBuildItem(cn, panacheFieldAccessEnhancer, modelClassNamesInternal));
+                        }
+                    }
                 }
             }
         }

--- a/extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheJpaEntityEnhancer.java
+++ b/extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheJpaEntityEnhancer.java
@@ -41,7 +41,7 @@ public class PanacheJpaEntityEnhancer extends PanacheEntityEnhancer<MetamodelInf
     private static final DotName DOTNAME_TRANSIENT = DotName.createSimple(Transient.class.getName());
 
     public PanacheJpaEntityEnhancer(IndexView index, List<PanacheMethodCustomizer> methodCustomizers) {
-        super(index, PanacheResourceProcessor.DOTNAME_PANACHE_ENTITY_BASE, methodCustomizers);
+        super(index, PanacheHibernateResourceProcessor.DOTNAME_PANACHE_ENTITY_BASE, methodCustomizers);
         modelInfo = new MetamodelInfo<>();
     }
 

--- a/extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheJpaRepositoryEnhancer.java
+++ b/extensions/panache/hibernate-orm-panache/deployment/src/main/java/io/quarkus/hibernate/orm/panache/deployment/PanacheJpaRepositoryEnhancer.java
@@ -17,7 +17,7 @@ public class PanacheJpaRepositoryEnhancer extends PanacheRepositoryEnhancer {
             .createSimple(PanacheRepositoryBase.class.getName());
 
     public PanacheJpaRepositoryEnhancer(IndexView index) {
-        super(index, PanacheResourceProcessor.DOTNAME_PANACHE_REPOSITORY_BASE);
+        super(index, PanacheHibernateResourceProcessor.DOTNAME_PANACHE_REPOSITORY_BASE);
     }
 
     @Override

--- a/extensions/panache/hibernate-orm-panache/runtime/pom.xml
+++ b/extensions/panache/hibernate-orm-panache/runtime/pom.xml
@@ -84,6 +84,12 @@
     <build>
         <plugins>
             <plugin>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <compilerArgument>-proc:none</compilerArgument>
+                </configuration>
+            </plugin>
+            <plugin>
                 <groupId>io.quarkus</groupId>
                 <artifactId>quarkus-bootstrap-maven-plugin</artifactId>
             </plugin>

--- a/extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheMongoEntityEnhancer.java
+++ b/extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheMongoEntityEnhancer.java
@@ -31,7 +31,7 @@ public class PanacheMongoEntityEnhancer extends PanacheEntityEnhancer<MetamodelI
     final Map<String, EntityModel> entities = new HashMap<>();
 
     public PanacheMongoEntityEnhancer(IndexView index, List<PanacheMethodCustomizer> methodCustomizers) {
-        super(index, PanacheResourceProcessor.DOTNAME_PANACHE_ENTITY_BASE, methodCustomizers);
+        super(index, PanacheMongoResourceProcessor.DOTNAME_PANACHE_ENTITY_BASE, methodCustomizers);
         modelInfo = new MetamodelInfo<>();
     }
 

--- a/extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheMongoRepositoryEnhancer.java
+++ b/extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheMongoRepositoryEnhancer.java
@@ -16,7 +16,7 @@ public class PanacheMongoRepositoryEnhancer extends PanacheRepositoryEnhancer {
     public final static DotName PANACHE_REPOSITORY_NAME = DotName.createSimple(PanacheMongoRepository.class.getName());
 
     public PanacheMongoRepositoryEnhancer(IndexView index) {
-        super(index, PanacheResourceProcessor.DOTNAME_PANACHE_REPOSITORY_BASE);
+        super(index, PanacheMongoResourceProcessor.DOTNAME_PANACHE_REPOSITORY_BASE);
     }
 
     @Override

--- a/extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheMongoResourceProcessor.java
+++ b/extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/PanacheMongoResourceProcessor.java
@@ -24,6 +24,8 @@ import org.jboss.jandex.Type;
 import io.quarkus.arc.deployment.BeanArchiveIndexBuildItem;
 import io.quarkus.arc.deployment.UnremovableBeanBuildItem;
 import io.quarkus.arc.deployment.ValidationPhaseBuildItem;
+import io.quarkus.bootstrap.classloading.ClassPathElement;
+import io.quarkus.bootstrap.classloading.QuarkusClassLoader;
 import io.quarkus.builder.BuildException;
 import io.quarkus.deployment.Capabilities;
 import io.quarkus.deployment.annotations.BuildProducer;
@@ -59,7 +61,7 @@ import io.quarkus.panache.common.deployment.PanacheMethodCustomizer;
 import io.quarkus.panache.common.deployment.PanacheMethodCustomizerBuildItem;
 import io.quarkus.panache.common.deployment.PanacheRepositoryEnhancer;
 
-public class PanacheResourceProcessor {
+public class PanacheMongoResourceProcessor {
     // blocking types
     static final DotName DOTNAME_PANACHE_REPOSITORY_BASE = DotName.createSimple(PanacheMongoRepositoryBase.class.getName());
     private static final DotName DOTNAME_PANACHE_REPOSITORY = DotName.createSimple(PanacheMongoRepository.class.getName());
@@ -81,6 +83,7 @@ public class PanacheResourceProcessor {
             .createSimple(ReactivePanacheMongoEntity.class.getName());
 
     private static final DotName DOTNAME_OBJECT_ID = DotName.createSimple(ObjectId.class.getName());
+    protected static final String META_INF_PANACHE_ARCHIVE_MARKER = "META-INF/panache-archive.marker";
 
     @BuildStep
     CapabilityBuildItem capability() {
@@ -116,7 +119,7 @@ public class PanacheResourceProcessor {
         Indexer indexer = new Indexer();
         Set<DotName> additionalIndex = new HashSet<>();
         IndexingUtil.indexClass(ObjectId.class.getName(), indexer, index.getIndex(), additionalIndex,
-                PanacheResourceProcessor.class.getClassLoader());
+                PanacheMongoResourceProcessor.class.getClassLoader());
         CompositeIndex compositeIndex = CompositeIndex.create(index.getIndex(), indexer.complete());
         Type type = Type.create(DOTNAME_OBJECT_ID, Type.Kind.CLASS);
         return new ReflectiveHierarchyBuildItem(type, compositeIndex);
@@ -192,8 +195,10 @@ public class PanacheResourceProcessor {
                 modelEnhancer.collectFields(classInfo);
         }
 
+        Set<String> modelClassNamesInternal = new HashSet<>();
         // iterate over all the entity classes
         for (String modelClass : modelClasses) {
+            modelClassNamesInternal.add(modelClass.replace(".", "/"));
             transformers.produce(new BytecodeTransformerBuildItem(modelClass, modelEnhancer));
 
             //register for reflection entity classes
@@ -209,10 +214,17 @@ public class PanacheResourceProcessor {
         if (!modelEnhancer.entities.isEmpty()) {
             PanacheFieldAccessEnhancer panacheFieldAccessEnhancer = new PanacheFieldAccessEnhancer(
                     modelEnhancer.getModelInfo());
-            for (ClassInfo classInfo : applicationIndex.getIndex().getKnownClasses()) {
-                String className = classInfo.name().toString();
-                if (!modelClasses.contains(className)) {
-                    transformers.produce(new BytecodeTransformerBuildItem(className, panacheFieldAccessEnhancer));
+            QuarkusClassLoader tccl = (QuarkusClassLoader) Thread.currentThread().getContextClassLoader();
+            List<ClassPathElement> archives = tccl.getElementsWithResource(META_INF_PANACHE_ARCHIVE_MARKER);
+            for (ClassPathElement i : archives) {
+                for (String res : i.getProvidedResources()) {
+                    if (res.endsWith(".class")) {
+                        String cn = res.replace("/", ".").substring(0, res.length() - 6);
+                        if (!modelClasses.contains(cn)) {
+                            transformers.produce(
+                                    new BytecodeTransformerBuildItem(cn, panacheFieldAccessEnhancer, modelClassNamesInternal));
+                        }
+                    }
                 }
             }
         }

--- a/extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/ReactivePanacheMongoEntityEnhancer.java
+++ b/extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/ReactivePanacheMongoEntityEnhancer.java
@@ -31,7 +31,7 @@ public class ReactivePanacheMongoEntityEnhancer extends PanacheEntityEnhancer<Me
     final Map<String, EntityModel> entities = new HashMap<>();
 
     public ReactivePanacheMongoEntityEnhancer(IndexView index, List<PanacheMethodCustomizer> methodCustomizers) {
-        super(index, PanacheResourceProcessor.DOTNAME_MUTINY_PANACHE_ENTITY_BASE, methodCustomizers);
+        super(index, PanacheMongoResourceProcessor.DOTNAME_MUTINY_PANACHE_ENTITY_BASE, methodCustomizers);
         modelInfo = new MetamodelInfo<>();
     }
 

--- a/extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/ReactivePanacheMongoRepositoryEnhancer.java
+++ b/extensions/panache/mongodb-panache/deployment/src/main/java/io/quarkus/mongodb/panache/deployment/ReactivePanacheMongoRepositoryEnhancer.java
@@ -17,7 +17,7 @@ public class ReactivePanacheMongoRepositoryEnhancer extends PanacheRepositoryEnh
     public final static DotName PANACHE_REPOSITORY_NAME = DotName.createSimple(ReactivePanacheMongoRepository.class.getName());
 
     public ReactivePanacheMongoRepositoryEnhancer(IndexView index) {
-        super(index, PanacheResourceProcessor.DOTNAME_MUTINY_PANACHE_REPOSITORY_BASE);
+        super(index, PanacheMongoResourceProcessor.DOTNAME_MUTINY_PANACHE_REPOSITORY_BASE);
     }
 
     @Override

--- a/extensions/panache/mongodb-panache/runtime/pom.xml
+++ b/extensions/panache/mongodb-panache/runtime/pom.xml
@@ -59,6 +59,12 @@
     <build>
         <plugins>
             <plugin>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <compilerArgument>-proc:none</compilerArgument>
+                </configuration>
+            </plugin>
+            <plugin>
                 <groupId>io.quarkus</groupId>
                 <artifactId>quarkus-bootstrap-maven-plugin</artifactId>
             </plugin>

--- a/extensions/panache/panache-common/runtime/pom.xml
+++ b/extensions/panache/panache-common/runtime/pom.xml
@@ -22,5 +22,15 @@
             <artifactId>quarkus-arc</artifactId>
         </dependency>
     </dependencies>
+    <build>
+        <plugins>
+            <plugin>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <compilerArgument>-proc:none</compilerArgument>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
 
 </project>

--- a/extensions/panache/panache-common/runtime/src/main/java/io/quarkus/panache/common/runtime/PanacheAnnotationProcessor.java
+++ b/extensions/panache/panache-common/runtime/src/main/java/io/quarkus/panache/common/runtime/PanacheAnnotationProcessor.java
@@ -1,0 +1,70 @@
+package io.quarkus.panache.common.runtime;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.Set;
+
+import javax.annotation.processing.Completion;
+import javax.annotation.processing.ProcessingEnvironment;
+import javax.annotation.processing.Processor;
+import javax.annotation.processing.RoundEnvironment;
+import javax.lang.model.SourceVersion;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.TypeElement;
+import javax.tools.FileObject;
+import javax.tools.StandardLocation;
+
+/**
+ * An annotation processor that is used to generate a marker file, to tell Quarkus that this archive has a dependency on
+ * Pananche entities, and therefore may need to be transformed for enhanced field access.
+ *
+ * This works because any archive that depends on Panache will have a transitive dependency on panache-common, which
+ * contains this processor. The Javac compiler will then run this processor and create the marker file.
+ */
+public class PanacheAnnotationProcessor implements Processor {
+
+    @Override
+    public Set<String> getSupportedOptions() {
+        return Collections.emptySet();
+    }
+
+    @Override
+    public Set<String> getSupportedAnnotationTypes() {
+        return Collections.singleton("*");
+    }
+
+    @Override
+    public SourceVersion getSupportedSourceVersion() {
+        return SourceVersion.latestSupported();
+    }
+
+    @Override
+    public void init(ProcessingEnvironment processingEnv) {
+        try {
+            FileObject marker = processingEnv.getFiler().createResource(StandardLocation.CLASS_OUTPUT, "",
+                    "META-INF/panache-archive.marker");
+            try (OutputStream out = marker.openOutputStream()) {
+                out.write(
+                        "This file is a marker, it exists to tell Quarkus that this archive has a dependency on Panache, and may need to be transformed at build time"
+                                .getBytes(StandardCharsets.UTF_8));
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Override
+    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
+        return false;
+    }
+
+    @Override
+    public Iterable<? extends Completion> getCompletions(Element element, AnnotationMirror annotation, ExecutableElement member,
+            String userText) {
+        return Collections.emptyList();
+    }
+}

--- a/extensions/panache/panache-common/runtime/src/main/resources/META-INF/services/javax.annotation.processing.Processor
+++ b/extensions/panache/panache-common/runtime/src/main/resources/META-INF/services/javax.annotation.processing.Processor
@@ -1,0 +1,1 @@
+io.quarkus.panache.common.runtime.PanacheAnnotationProcessor

--- a/extensions/panache/panache-mock/pom.xml
+++ b/extensions/panache/panache-mock/pom.xml
@@ -34,4 +34,14 @@
             <artifactId>mockito-core</artifactId>
         </dependency>
     </dependencies>
+    <build>
+        <plugins>
+            <plugin>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <compilerArgument>-proc:none</compilerArgument>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
 </project>

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/CuratedApplication.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/CuratedApplication.java
@@ -21,6 +21,7 @@ import java.util.function.BiConsumer;
 import java.util.function.BiFunction;
 import java.util.function.Consumer;
 import java.util.function.Function;
+import java.util.function.Predicate;
 import org.objectweb.asm.ClassVisitor;
 
 /**
@@ -280,10 +281,12 @@ public class CuratedApplication implements Serializable, Closeable {
 
     public QuarkusClassLoader createRuntimeClassLoader(QuarkusClassLoader loader,
             Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> bytecodeTransformers,
+            Map<String, Predicate<byte[]>> transformerPredicates,
             ClassLoader deploymentClassLoader, Map<String, byte[]> resources) {
         QuarkusClassLoader.Builder builder = QuarkusClassLoader.builder("Quarkus Runtime ClassLoader",
                 loader, false)
                 .setAggregateParentResources(true);
+        builder.setTransformerPredicates(transformerPredicates);
         builder.setTransformerClassLoader(deploymentClassLoader);
         if (quarkusBootstrap.getApplicationRoot() != null) {
             builder.addElement(ClassPathElement.fromPath(quarkusBootstrap.getApplicationRoot()));

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java
@@ -8,6 +8,7 @@ import java.lang.reflect.InvocationTargetException;
 import java.net.URL;
 import java.security.ProtectionDomain;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.Enumeration;
 import java.util.HashMap;
@@ -19,6 +20,7 @@ import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 import java.util.function.BiFunction;
+import java.util.function.Predicate;
 import java.util.jar.Attributes;
 import java.util.jar.Manifest;
 import org.jboss.logging.Logger;
@@ -71,6 +73,7 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
     private volatile MemoryClassPathElement resettableElement;
 
     private volatile Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> bytecodeTransformers;
+    private volatile Map<String, Predicate<byte[]>> transformerPredicates;
     private volatile ClassLoader transformerClassLoader;
     private volatile ClassLoaderState state;
 
@@ -99,6 +102,7 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
         this.name = builder.name;
         this.elements = builder.elements;
         this.bytecodeTransformers = builder.bytecodeTransformers;
+        this.transformerPredicates = builder.transformerPredicates;
         this.bannedElements = builder.bannedElements;
         this.parentFirstElements = builder.parentFirstElements;
         this.lesserPriorityElements = builder.lesserPriorityElements;
@@ -126,12 +130,14 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
 
     public void reset(Map<String, byte[]> resources,
             Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> bytecodeTransformers,
+            Map<String, Predicate<byte[]>> transformerPredicates,
             ClassLoader transformerClassLoader) {
         if (resettableElement == null) {
             throw new IllegalStateException("Classloader is not resettable");
         }
         this.transformerClassLoader = transformerClassLoader;
         synchronized (this) {
+            this.transformerPredicates = transformerPredicates;
             resettableElement.reset(resources);
             this.bytecodeTransformers = bytecodeTransformers;
             state = null;
@@ -344,8 +350,9 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
                     if (classPathElementResource != null) { //can happen if the class loader was closed
                         byte[] data = classPathElementResource.getData();
                         List<BiFunction<String, ClassVisitor, ClassVisitor>> transformers = bytecodeTransformers.get(name);
+                        Map<String, Predicate<byte[]>> constPoolEntries = transformerPredicates;
                         if (transformers != null) {
-                            data = handleTransform(name, data, transformers);
+                            data = handleTransform(name, data, transformers, constPoolEntries);
                         }
                         definePackage(name, classPathElement);
                         return defineClass(name, data, 0, data.length,
@@ -391,6 +398,19 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
         }
     }
 
+    public List<ClassPathElement> getElementsWithResource(String name) {
+        List<ClassPathElement> ret = new ArrayList<>();
+        if (parent instanceof QuarkusClassLoader) {
+            ret.addAll(((QuarkusClassLoader) parent).getElementsWithResource(name));
+        }
+        ClassPathElement[] classPathElements = getState().loadableResources.get(name);
+        if (classPathElements == null) {
+            return ret;
+        }
+        ret.addAll(Arrays.asList(classPathElements));
+        return ret;
+    }
+
     private String getPackageNameFromClassName(String className) {
         final int index = className.lastIndexOf('.');
         if (index == -1) {
@@ -403,7 +423,13 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
     }
 
     private byte[] handleTransform(String name, byte[] bytes,
-            List<BiFunction<String, ClassVisitor, ClassVisitor>> transformers) {
+            List<BiFunction<String, ClassVisitor, ClassVisitor>> transformers,
+            Map<String, Predicate<byte[]>> transformerPredicates) {
+        if (transformerPredicates.containsKey(name)) {
+            if (!transformerPredicates.get(name).test(bytes)) {
+                return bytes;
+            }
+        }
         ClassReader cr = new ClassReader(bytes);
         ClassWriter writer = new ClassWriter(cr, ClassWriter.COMPUTE_FRAMES | ClassWriter.COMPUTE_MAXS) {
 
@@ -451,6 +477,7 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
         final List<ClassPathElement> bannedElements = new ArrayList<>();
         final List<ClassPathElement> parentFirstElements = new ArrayList<>();
         final List<ClassPathElement> lesserPriorityElements = new ArrayList<>();
+        Map<String, Predicate<byte[]>> transformerPredicates = Collections.emptyMap();
         Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> bytecodeTransformers = Collections.emptyMap();
         final boolean parentFirst;
         MemoryClassPathElement resettableElement;
@@ -561,6 +588,15 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
             } else {
                 this.bytecodeTransformers = bytecodeTransformers;
             }
+        }
+
+        /**
+         * Sets const pool entries that need to be present for a class to be transformed. This optimisation
+         * allows for transformation to be skipped if the scanner detects that no transformation is required.
+         */
+        public Builder setTransformerPredicates(Map<String, Predicate<byte[]>> transformerPredicates) {
+            this.transformerPredicates = transformerPredicates;
+            return this;
         }
 
         /**

